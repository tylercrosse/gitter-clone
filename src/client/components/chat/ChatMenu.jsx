import React from 'react';

export const ChatMenu = () => (
  <aside className="chat-menu">
    <nav>
      <ul>
        <li>G</li>
        <li>-</li>
      </ul>
    </nav>
  </aside>
);

export default ChatMenu;
