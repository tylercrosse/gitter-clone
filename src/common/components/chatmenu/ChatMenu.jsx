import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import {
  openCreateRoomModal,
  closeModal,
  openConvosPanel,
  openDirectMessagesPanel,
  closePanel,
} from '../../ducks/ui';
import { createConvo,
fetchConvos } from '../../ducks/convos';
import convosSelector from '../../selectors/convosSelector';
import directConvosSelector from '../../selectors/directConvosSelector';
import GitterCloneIcon from '../svgs/GitterCloneIcon';
import ConvosIcon from '../svgs/ConvosIcon';
import DirectIcon from '../svgs/DirectIcon';
import Panel from './Panel';

export class ChatMenu extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      active: false
    };
    this.handleMouseLeave = this.handleMouseLeave.bind(this);
  }
  componentDidMount() {
    this.props.fetchConvos();
  }
  handleMouseLeave(e) {
    /* istanbul ignore next: ignore leaving to minibar */
    if (e.clientX < 74 || this.props.modalIsOpen) return;
    this.props.closePanel();
  }
  render() {
    return (
      <aside className="chat-menu">
        <section className="chat-menu-container">
          <nav className="minibar">
            <div className="minibar-inner">
              <ul>
                <li className="minibar-convos">
                  <Link to={'/'} className="minibar-button">
                    <GitterCloneIcon classes="minibar-icon icon-gitter-clone" />
                  </Link>
                </li>
                <li className="minibar-all">
                  <button
                    className="minibar-button minibar-button-search"
                    onClick={this.props.openConvosPanel}
                  >
                    <ConvosIcon classes="minibar-icon icon-convo" />
                  </button>
                </li>
                {this.props.user.loggedIn &&
                  <li className="minibar-direct">
                    <button
                      className="minibar-button minibar-button-search"
                      onClick={this.props.openDirectMessagesPanel}
                    >
                      <DirectIcon classes="minibar-icon icon-direct" />
                    </button>
                  </li>}
              </ul>
            </div>
          </nav>
          <Panel
            onMouseLeave={this.handleMouseLeave}
            active={this.props.panel.open}
            title={this.props.panel.title}
            inner={this.props.panel.inner}
            {...this.props}
          />
        </section>
      </aside>
    );
  }
}

ChatMenu.propTypes = {
  panel: PropTypes.shape({
    open: PropTypes.bool.isRequired,
    title: PropTypes.string.isRequired,
    inner: PropTypes.string.isRequired,
  }).isRequired,
  user: PropTypes.shape({
    loggedIn: PropTypes.bool.isRequired,
  }).isRequired,
  modalIsOpen: PropTypes.bool.isRequired,
  openConvosPanel: PropTypes.func.isRequired,
  openDirectMessagesPanel: PropTypes.func.isRequired,
  closePanel: PropTypes.func.isRequired,
  fetchConvos: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  modalIsOpen: state.ui.modalIsOpen.createRoom,
  user: state.user,
  panel: state.ui.panel,
  convos: convosSelector(state),
  directConvos: directConvosSelector(state), //FIXME
});

export default connect(mapStateToProps, {
  openCreateRoomModal,
  closeModal,
  createConvo,
  fetchConvos,
  openConvosPanel,
  openDirectMessagesPanel,
  closePanel
})(ChatMenu);
