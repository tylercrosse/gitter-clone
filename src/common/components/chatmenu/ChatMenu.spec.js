import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { ChatMenu, mapStateToProps } from './ChatMenu';

const setup = () => {
  const props = {
    modalIsOpen: false,
    openCreateRoomModal: jest.fn(),
    closeModal: jest.fn(),
    addConvo: jest.fn(),
    fetchConvos: jest.fn(),
    user: {
      loggedIn: true
    },
    convos: [
      {
        id: '9ebb6983-50cf-4226-ab4b-1ebd12db80a3',
        name: 'chat',
        direct: false,
        createdAt: '2017-06-20T21:21:24.752Z',
        updatedAt: '2017-06-20T21:21:24.752Z'
      },
      {
        id: 'f55f180f-3974-41ab-a521-931d52101914',
        name: 'react',
        direct: false,
        createdAt: '2017-06-20T21:21:29.312Z',
        updatedAt: '2017-06-20T21:21:29.312Z'
      }
    ],
    panel: {
      open: false,
      title: 'chat2',
      inner: 'chat2',
    },
    openConvosPanel: jest.fn(),
    openDirectMessagesPanel: jest.fn(),
    closePanel: jest.fn(),
  };

  const component = <ChatMenu {...props} />;

  return {
    props,
    component
  };
};

describe('<ChatMenu />', () => {
  it('should render correctly', () => {
    const { component } = setup();
    const renderedComponent = renderer.create(component);
    const tree = renderedComponent.toJSON();
    expect(tree).toMatchSnapshot();
  });

  xit('should toggle active state on menu button click', () => {
    const { component } = setup();
    const wrapper = mount(component);
    expect(wrapper.state().active).toBeFalsy();
    wrapper.find('.minibar-button-convos').simulate('click');
    expect(wrapper.state().active).toBeTruthy();
    wrapper.find('.minibar-button-convos').simulate('click');
    expect(wrapper.state().active).toBeFalsy();
  });

  xit('should toggle active state onMouseLeave', () => {
    const { component } = setup();
    const wrapper = mount(component);
    expect(wrapper.state().active).toBeFalsy();
    wrapper.find('.minibar-button-convos').simulate('click');
    expect(wrapper.state().active).toBeTruthy();
    // trouble simulating clientX on synthetic mouseLeave event
    wrapper.instance().handleMouseLeave({clientX: 400});
    expect(wrapper.state().active).toBeFalsy();
  });

  xit('should apply class "active" when state is active', () => {
    const { component } = setup();
    const wrapper = mount(component);
    expect(wrapper.state().active).toBeFalsy();
    expect(wrapper.find('.chat-menu-panel').hasClass('active'))
      .toBeFalsy();
    wrapper.find('.minibar-button-convos').simulate('click');
    expect(wrapper.state().active).toBeTruthy();
    expect(wrapper.find('.chat-menu-panel').hasClass('active'))
      .toBeTruthy();
  });

  it('should receive the correct props from state', () => {
    const { props } = setup();
    const ui = {
      modalIsOpen: { createRoom: false },
      panel: props.panel
    };
    expect(mapStateToProps({...props, ui}))
      .toEqual({
        modalIsOpen: props.modalIsOpen,
        user: props.user,
        panel: props.panel,
        convos: props.convos,
        directConvos: []
      });
  });
});
