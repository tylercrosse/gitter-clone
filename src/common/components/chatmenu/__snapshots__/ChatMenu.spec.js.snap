// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatMenu /> should render correctly 1`] = `
<aside
  className="chat-menu"
>
  <section
    className="chat-menu-container"
  >
    <nav
      className="minibar"
    >
      <div
        className="minibar-inner"
      >
        <ul>
          <li
            className="minibar-convos"
          >
            <a
              className="minibar-button"
              onClick={[Function]}
              style={Object {}}
            >
              <svg
                className="minibar-icon icon-gitter-clone"
                fill="#ba9cc3"
                height="22px"
                version="1.1"
                viewBox="0 0 22.43 32.04"
                width="15px"
                x="0px"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <defs />
                <g>
                  <rect
                    height="24.03"
                    width="3.2"
                    y="4.81"
                  />
                  <rect
                    height="5.61"
                    width="3.2"
                    x="6.41"
                    y="1.6"
                  />
                  <rect
                    height="8.81"
                    width="3.2"
                    x="19.22"
                  />
                  <rect
                    height="7.21"
                    width="3.2"
                    x="19.22"
                    y="24.83"
                  />
                  <rect
                    height="16.02"
                    width="3.2"
                    x="6.41"
                    y="8.81"
                  />
                  <rect
                    height="5.61"
                    width="3.2"
                    x="6.41"
                    y="26.43"
                  />
                  <rect
                    height="4"
                    width="3.2"
                    x="12.81"
                    y="28.03"
                  />
                  <rect
                    height="20.82"
                    width="3.2"
                    x="12.81"
                    y="5.61"
                  />
                  <rect
                    height="4"
                    width="3.2"
                    x="12.81"
                  />
                  <rect
                    height="12.81"
                    width="3.2"
                    x="19.22"
                    y="10.41"
                  />
                </g>
              </svg>
            </a>
          </li>
          <li
            className="minibar-all"
          >
            <button
              className="minibar-button minibar-button-convos"
              onClick={[Function]}
            >
              <svg
                className="minibar-icon icon-convo"
                height="22"
                viewBox="0 0 30.37 26.65"
                width="22"
                x="0px"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <path
                  d="M15.81,4.53C8,4.53,1.63,9,1.63,14.58c0,3.35,2.32,6.32,5.88,8.15C7,26.11,2.82,29.44,3.67,29.16a31.32,31.32,0,0,0,9.74-4.68,20,20,0,0,0,2.4.14c7.84,0,14.19-4.5,14.19-10.05S23.65,4.53,15.81,4.53Z"
                  transform="translate(-0.63 -3.53)"
                />
              </svg>
            </button>
          </li>
          <li
            className="minibar-direct"
          >
            <button
              className="minibar-button minibar-button-direct"
              onClick={[Function]}
            >
              <svg
                className="minibar-icon icon-direct"
                height="22"
                viewBox="0 0 24 25.22"
                width="22"
                x="0px"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <ellipse
                  cx="12"
                  cy="6.48"
                  rx="4.59"
                  ry="5.48"
                />
                <path
                  d="M27,29.21C27,23.57,22.08,19,16,19S5,23.57,5,29.21"
                  transform="translate(-4 -3.99)"
                />
              </svg>
            </button>
          </li>
        </ul>
      </div>
    </nav>
    <section
      className="chat-menu-panel "
      onMouseLeave={[Function]}
    >
      <div
        className="brand-container"
      >
        Gitter Clone
      </div>
      <header
        className="panel-header"
      >
        <div
          className="panel-header-convos"
        >
          <h2>
            chat2
          </h2>
        </div>
      </header>
    </section>
  </section>
</aside>
`;
